Fix package detection info in DiscoveredPackages #1099

AyanSinhaMahapatra opened the pull request:
Reference: https://github.com/nexB/scancode.io/issues/1099

tdruez commented:
@AyanSinhaMahapatra 

> Should we show this in the DiscoveredPackageListView (similar to how we show related fields in DiscoveredDependencyListView)?

This does not seems necessary. The space is limited in list views and the detection info are not essentials.

> In DiscoveredPackageListView and DiscoveredPackageDetailsView should we link to the CodebaseResourceDetailsView for each datafile path (similar to what we do in dependencies)?

Sure, why not? Be conscious about the db queries though, but I guess we can generate the links just using the path without any extra queries.

> Here we have an extra codebase/ appearing in packages created by scan_codebase_packages pipeline due to Debian package is not extracted recursively #1081 (comment), is it okay to remove this specifically if it appears in front of the path?

This sounds like a hack, it would be better to store the proper path in the db than presenting altered data in the UI.

> How do we show these fields in DiscoveredPackageDetailsView, since these are lists now, and with links, in case of datafile_paths?

What about a new "Detection" tab, similar to the one on the CodebaseResourceDetailsView ?

----

Also, the migration includes a file type change, from CharField to JSONField for `datasource_ids`, yet it does not include any data migration.
Are you expecting any issues from this? Have you tried to run this on a real dataset?

AyanSinhaMahapatra commented:
> Sure, why not? Be conscious about the db queries though, but I guess we can generate the links just using the path without any extra queries.

Yes, we're not doing any queries here, just creating the links in the templates.

> This sounds like a hack, it would be better to store the proper path in the db than presenting altered data in the UI.

Doing this on the package creation side, and storing the right path now.

> What about a new "Detection" tab, similar to the one on the CodebaseResourceDetailsView ?

This sounds good, adding this.

> Also, the migration includes a file type change, from CharField to JSONField for datasource_ids, yet it does not include any data migration. Are you expecting any issues from this? Have you tried to run this on a real dataset?

So we were not storing any datasource IDs previously, due to the bug linked, and so my assumption was we don't have any datasource IDs for any packages stored from previous scans. This was because there was a attribute mismatch and we were ignoring this data while importing from PackageData/creating a DiscoveredPackage. But when I'm checking the DiscoveredPackage objects for a SCIO instance, there were some Packages with these datasource_id fields populated for the new inspect_packages pipeline, or anywhere we had Packages created out of PackageData objects, so I added a migration for this now.

tdruez commented on scanpipe/pipes/__init__.py:
```diff
@@ -155,6 +155,16 @@ def _clean_package_data(package_data):
+    # Strip leading "codebase/" to make path compatible with
+    # paths stored in resource database
+    cleaned_datafile_paths = []
+    for path in package_data.get("datafile_paths", []):
+        cleaned_path = path
+        if path.startswith("codebase/"):
+            cleaned_path = path[len("codebase/") :]
+        cleaned_datafile_paths.append(cleaned_path)
```
What about: 
`cleaned_datafile_paths = [path.removeprefix("codebase/") for path in package_data.get("datafile_paths", [])]`

AyanSinhaMahapatra replied:
Using `path.removeprefix` is much cleaner, thanks! Added this.

tdruez reviewed (changes requested):
@AyanSinhaMahapatra Could you add a changelog entry to explain those changes?
Also, have you tried the data migrations on a real dataset?

AyanSinhaMahapatra commented:
@tdruez I've added your feedback and this is ready to review again.

> Could you add a changelog entry to explain those changes?

Added this now.

> Also, have you tried the data migrations on a real dataset?

Yes, I had created a test dataset with pipeline runs of all pipeline types and different kinds of discovered packages created from those pipelines (including some DiscoveredPackages with `datasource_id` values from non-empty inspect_packages pipelines) and the data migration worked nicely on this without errors. See output below:
```
-> Apply database migrations
bin/python manage.py migrate
Operations to perform:
  Apply all migrations: admin, auth, authtoken, contenttypes, django_rq, scanpipe, sessions, taggit
Running migrations:
  Applying scanpipe.0054_discoveredpackage_datafile_paths...
Compute datasource_ids for 22 packages.
Updating DB objects...
 OK
-> Run the test suite
bin/python manage.py test --noinput
Found 526 test(s).
Creating test database for alias 'default'...

Compute datasource_ids for 0 packages.
Updating DB objects...
System check identified no issues (0 silenced).
.......................................................................................................................................................................................x......................................................................................................................................................................................................................................................................................................................................................
----------------------------------------------------------------------
Ran 526 tests in 100.089s

OK (expected failures=1)
Destroying test database for alias 'default'...
```

I've also merged latest main to resolve the conflicts and renamed the migration accordingly.
